stakingModel/contracts/InterestDistribution.sol

```diff
@@ -0,0 +1,193 @@
+     */
+    struct Staker {
+        uint256 stakedToken;
+        uint256 weightedStake; // stored with 2 precision points ie., 0.82 => 82
+        uint256 lastStake;
+        uint256 withdrawnToDate;
+        uint256 avgYieldRatePerToken;
+    }
+
+    // 10^18
+    uint256 constant DECIMAL1e18 = 10 ** 18;
+
+    /**
+      * @dev Updates InterestData and Staker data while staking.
+      * 
+      * @param _interestData           Interest data
+      * @param _staker                 Staker's address
+      * @param _stake                  Amount of stake
+      * @param _donationPer            Percentage will to donate.
+      * @param _dailyIntrest           Withdrawn interest by individual staker so far.
+      * @param _grandTotalStaked       Average yielding rate per token
+      *
+    */
+    function stakeCalculation(
+      InterestData storage _interestData, 
+      address _staker,
+      uint256 _stake, 
+      uint256 _donationPer,
+      uint256 _dailyIntrest,
+      uint256 _grandTotalStaked
+      ) 
+    internal 
+    {
+      _interestData.grandTotalStaked = _interestData.grandTotalStaked.add(_stake);
+      uint currentStake = _interestData.stakers[_staker].stakedToken;
+      Staker storage _stakerData = _interestData.stakers[_staker];
+      _stakerData.stakedToken = currentStake.add(_stake);
+      _stakerData.weightedStake = (_stakerData.weightedStake.mul(currentStake).add(_stake.mul(uint(100).sub(_donationPer)))).div(_stakerData.stakedToken);
+      _stakerData.lastStake = block.number;
+      uint _accumulatedYieldPerToken = getAccumulatedYieldPerToken(_dailyIntrest, _grandTotalStaked);
+      _stakerData.avgYieldRatePerToken = _stakerData.avgYieldRatePerToken.add(getAvgYieldRatePerToken(_accumulatedYieldPerToken, _stake, _donationPer, _stakerData.stakedToken));
+    }
+
+    /**
+      * @dev Updates InterestData and Staker data while withdrawing stake.
+      * 
+      * @param _interestData           Interest data
+      * @param _staker                 Staker address
+      * @param _amount                 Amount of stake to withdraw
+      *
+    */
+    function withdrawStakeAndInterest(
+      InterestData storage _interestData, 
+      address _staker, 
+      uint256 _amount
+      ) 
+    internal 
+    {
+      _interestData.grandTotalStaked = _interestData.grandTotalStaked.sub(_amount);
+      Staker storage _stakerData = _interestData.stakers[_staker];
+      _stakerData.stakedToken = _stakerData.stakedToken.sub(_amount);
+    }
+
+    /**
+      * @dev Updates interestAccrued and accumulatedYieldPerToken.
+      * 
+      * @param _interestData           Interest data
+      * @param _interest               Newly accrued interest
+      * @param _grandTotalStaked       Average yielding rate per token
+      *
+    */
+    function addInterest(InterestData storage _interestData, uint256 _interest, uint256 _grandTotalStaked) internal {
```

Comment:
1.  instead of "add" prefix functions with "update" (change in other methods were relevant)
2. why is grandTotalSTaked and input? the _interestData holds it

Reply:
Done.